devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,256 @@
+    >>> u = Function(name='u', grid=grid, space_order=2, coefficients='symbolic')
+    >>> x, y = grid.dimensions
+
+    Now define some partial d/dx FD coefficients of the Function u:
+
+    >>> u_x_coeffs = Coefficient(1, u, x, np.array([-0.6, 0.1, 0.6]))
+
+    And some partial d^2/dy^2 FD coefficients:
+
+    >>> u_y2_coeffs = Coefficient(2, u, y, np.array([0.0, 0.0, 0.0]))
+
+    """
+
+    def __init__(self, deriv_order, function, dimension, coefficients):
+
+        self.check_input(deriv_order, function, dimension, coefficients)
+
+        # Ensure the given set of coefficients is the correct length
+        if dimension.is_Time:
+            if len(coefficients)-1 != function.time_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+        elif dimension.is_Space:
+            if len(coefficients)-1 != function.space_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+
+        self._deriv_order = deriv_order
+        self._function = function
+        self._dimension = dimension
+        self._coefficients = coefficients
+
+    @property
+    def deriv_order(self):
+        """The derivative order."""
+        return self._deriv_order
+
+    @property
+    def function(self):
+        """The function to which the coefficients belong."""
+        return self._function
+
+    @property
+    def dimension(self):
+        """The dimension to which the coefficients will be applied."""
+        return self._dimension
+
+    @property
+    def coefficients(self):
+        """The set of coefficients."""
+        return self._coefficients
+
+    def check_input(self, deriv_order, function, dimension, coefficients):
+        if not isinstance(deriv_order, int):
+            raise TypeError("Derivative order must be an integer")
+        # NOTE: Can potentially be tidied up following the implementation
+        # of lazy evaluation.
+        try:
+            if not function.is_Function:
+                raise TypeError("Object is not of type Function")
+        except:
+            raise TypeError("Object is not of type Function")
+        try:
+            if not dimension.is_Dimension:
+                raise TypeError("Coefficients must be attached to a valid dimension")
+        except:
+            raise TypeError("Coefficients must be attached to a valid dimension")
+        # Currently only numpy arrays are accepted here.
+        # Functionality will be expanded in the near future.
+        if not isinstance(coefficients, np.ndarray):
+            raise NotImplementedError
+        return
+
+
+class CoefficientRules(object):
+    """
+    Devito class to convert Coefficient objects into replacent rules
+    to be applied when constructing a Devito Eq.
+
+    Example
+    -------
+    Assume we have created the 'u_x_coeffs' Coefficient object
```

Comment:
OK, this is quite important. I'd explicitly recreate ``u_x_coeffs`` ...

... and I'd add this file to ``test_docstrings.py``, [here](https://github.com/opesci/devito/blob/master/tests/test_docstrings.py#L27) , so that CI automatically tests the correctness of all these example code snippets and we're sure they are never out of date. 

Note: I was suggesting to recreate ``u_x_coeffs`` explicitly otherwise the test will fail complaining it couldn't find ``u_x_coeffs``

That's why in another comment I was suggesting to also show the output of some of these variables , although perhaps as you said in some cases it's not that meaningful

Reply:
Updated + `coefficients.py` added to `test_docstrings.py`.

Quick note: printing u_x_coeffs returns things such as `<devito.finite_differences.coefficients.Coefficient object at 0x7f0fa3f4da20>` which is why the output is not currently being printed (and no output will be printed by default owing to the '=').